### PR TITLE
schedulers: evict leader support multiple targets

### PR DESCRIPTION
**Pull request by MrCroxx**
### What problem does this PR solve?

Part of https://github.com/tikv/tikv/issues/10602.

close [#4229](https://github.com/tikv/pd/issues/4229)

### What is changed and how it works?

Evict leader scheduler picks all valid stores, let TiKV decide the best store to transfer leader to.

### Check List

Tests

- [x] Unit test
- [x] Integration test
- [x] Manual test, see [tikv#10602](tikv/tikv#10602).

Side effects

- Evict leader may lead to leader unbalanced when some store is slow.

Related changes

- [kvproto#822](https://github.com/pingcap/kvproto/pull/822)
- [tikv#11063](https://github.com/tikv/tikv/pull/11063)

### Release note

```release-note
Evict leader automatically pick the fastest store to transfer leader to.
```

Signed-off-by: MrCroxx <mrcroxx@outlook.com>

### CODE REVIEW

**Comment by HunDunDM** [comment]
* Maybe using another name would be better than `TransferLeaderV2`, also in `kvproto`, it is a special step, not an enhancement of `TransferLeader`. (Refer to `ChangePeer` and `ChangePeerV2`)
* Refer to `versioninfo.JointConsensus` to implement backward compatibility and ensure that the `TransferLeader` is sent to TiKVs that do not support this feature.

**Comment by MrCroxx** [comment]
> * Maybe using another name would be better than `TransferLeaderV2`, also in `kvproto`, it is a special step, not an enhancement of `TransferLeader`. (Refer to `ChangePeer` and `ChangePeerV2`)

From PD's view, the new evict leader implementation is not that similar with transfer leader, but from TiKV's view they are. IMO, reuse the origin `TransferLeader` proto with tiny modifications (`peer` => `peers`) is just okay? What do yo think? @BusyJay @NingLin-P

**Comment by nolouch** [comment]
I wonder even the evict leader picks the best one, another scheduler like `balance-leader` will schedule the leader to the other store in order to maintain the leader number is balanced.

**Comment by MrCroxx** [comment]
> I wonder even the evict leader picks the best one, another scheduler like `balance-leader` will schedule the leader to the other store in order to maintain the leader number is balanced.

@nolouch IMO, this modification is only for speeding up evict leader ops and making the store can be brought offline faster. It's fine that the transfered leaders need to be balanced smoothly later.

**Comment by MrCroxx** [comment]
Let me summarize the conversations above.

1. Reuse the origin `TransferLeader` proto message by replacing singular field `peer` with repeatable field `peers`.
2. Add an config in PD to decide whether to enable the multi-target leader evicting feature.
3. Check versioninfo to decide whether to enable the feature. (or simply use random order in `peers` field to simplify the compatible problem?)

@HunDunDM Any updates?

**Comment by BusyJay** [comment]
>     1. Reuse the origin `TransferLeader` proto message by replacing singular field `peer` with repeatable field `peers`.

Unfortunately, according to https://github.com/tikv/pd/pull/4214#issuecomment-946553963 it's not backward compatible, hence should use a new field instead.

**Comment by nolouch** [comment]
Could you test the branch to show how many speeds improve？

File: server/schedule/operator/create_operator.go [hunk 1]
```diff
@@ -73,6 +74,19 @@ func CreateForceTransferLeaderOperator(desc string, cluster opt.Cluster, region
 		Build(kind)
 }
 
+// CreateEvictLeaderOperator creates an operator that transfers the leader from a source store to one of the given target stores
+// or the given target store when incompatible.
+func CreateEvictLeaderOperator(desc string, cluster opt.Cluster, region *core.RegionInfo, sourceStoreID uint64, targetStoreID uint64, targetStoreIDs []uint64, kind OpKind) (*Operator, error) {
```
Comment (rleungx) [inline server/schedule/operator/create_operator.go:79]: Can we use a more general name since it could be used in other scenes?

Comment (MrCroxx) [reply server/schedule/operator/create_operator.go:79]: Emm, in my first edition implementation, I directly modified `TransferLeader` to support multi targets. But then we found only evict leader needs it, and other ops only needs one target. So I created a seperate `EvictLeader`

Comment (MrCroxx) [reply server/schedule/operator/create_operator.go:79]: @rleungx Any updates about the naming? I've not come up with a good idea yet.

Comment (rleungx) [reply server/schedule/operator/create_operator.go:79]: No, but it needs to be general. ^ ^

File: server/schedule/operator/step.go [hunk 2]
```diff
@@ -82,6 +82,70 @@ func (tl TransferLeader) Influence(opInfluence OpInfluence, region *core.RegionI
 	to.LeaderCount++
 }
 
+// EvictLeader is an OpStep that transfers a region's leader to one of the targets.
+type EvictLeader struct {
+	FromStore uint64
+	ToStore   uint64
```
Comment (rleungx) [inline server/schedule/operator/step.go:88]: Why do we still need it?

Comment (MrCroxx) [reply server/schedule/operator/step.go:88]: For compatible, [ref](https://github.com/tikv/pd/pull/4214#issuecomment-947587897).

Comment (rleungx) [reply server/schedule/operator/step.go:88]: I think we don't need this field here, it can be appended when we send it to TiKV?

**Comment by MrCroxx** [comment]
Duplicated #4401 .

**Pull request closed without merging**